Report tab; Bootstrap3 compatible and reported-user-detail button

Zauberstuhl opened the pull request:
![untitled](https://cloud.githubusercontent.com/assets/652428/9663191/7aecc248-5263-11e5-992e-829a6a708c35.png)

b3 hasn't triggered that.. more the new button.

It will search the guid via (admin) user_search.. This can be useful if the podmin
wants to lock, unlock or delete the account responsible for the reported post.

Commit:
Report tab; bootstrap3 compatible, add user detail btn

The reported_user_details button will redirect you
to admin user_search. This can be useful if the podmin
wants to lock, unlock or delete the account which is
responsible for the reported post

Signed-off-by: Lukas Matt <lukas@zauberstuhl.de>

Commit:
Add get_reported_guid spec for report helper

* two new methods in report model reported_author and item
* merge deletion methods in report model

Signed-off-by: Lukas Matt <lukas@zauberstuhl.de>

svbergerem commented on app/assets/stylesheets/report.scss:
```diff
@@ -1,21 +1,9 @@
 #reports {
   padding-top: 2em;
```
I'd remove this line to move the headline to the top of the page.

svbergerem commented:
UI review done. Someone else™ has to do a RoR review.

jhass commented on app/views/report/index.html.haml:
```diff
@@ -3,26 +3,30 @@
-              = button_to t('report.review_link'), report_path(r.id, :type => r.item_type),
-                :class => "btn btn-info btn-small",
+          .panel.panel-default
+            - username = User.find_by_id(r.user_id).username
```
`report.user.username` didn't work?

jhass commented on app/views/report/index.html.haml:
```diff
@@ -3,26 +3,30 @@
+                = t("report.reason_label", text: r.text)
+
+              = button_to t("report.reported_user_details"),
+                user_search_path(admins_controller_user_search: {guid: get_reported_guid(r.item_id,
```
Formatting seems a bit odd here, how about:

```haml
              = button_to t("report.reported_user_details"),
                user_search_path(admins_controller_user_search: {
                  guid: get_reported_guid(r.item_id, r.item_type)
                }), class: "btn pull-left btn-info btn-small",

jhass commented on app/helpers/report_helper.rb:
```diff
@@ -3,15 +3,24 @@
 #   the COPYRIGHT file.
 
 module ReportHelper
+  def get_reported_guid(id, type)
```
Leave off `get_`, `set_` (replace with `=` suffix) and `is_` (replace with `?` suffix) prefixes in Ruby. Also this should be fine with being a method on the Report model, becoming `report.reported_author.guid` in the view.

jhass commented on app/helpers/report_helper.rb:
```diff
@@ -3,15 +3,24 @@
       # cross-site-scripting we have to escape html
-      raw t('report.comment_label', data: link_to(h(comment_message(comment)), post_path(comment.post.id, anchor: comment.guid)))
+      raw t("report.comment_label", data: link_to(
+        h(comment_message(comment)), post_path(comment.post.id, anchor: comment.guid)))
```
While touching this clean it up to `comment.post_id`

jhass commented on app/helpers/report_helper.rb:
```diff
@@ -3,15 +3,24 @@
     else
-      raw t('report.not_found')
+      raw t("report.not_found")
     end
```
Also while touching this I feel like we can clean it up with a method or association on the Report model or two so it becomes

```rb
case report.item
when Post
  # ...
when Comment
  # ...
else
  # not found
end

jhass commented on app/views/report/index.html.haml:
```diff
@@ -3,26 +3,30 @@
-              = button_to t('report.delete_link'), report_path(r.id, :type => r.item_type),
-                :data => { :confirm => t('report.confirm_deletion') },
-                :class => "btn btn-danger btn-small",
+              = button_to t("report.delete_link"), report_path(r.id, type: r.item_type),
```
Maybe we can even get rid of passing the type here, by looking it up in the report fetched via its id in the controller.

Commit:
Check if post or comment exist while rendering

It is possible that the item was deleted
during the time it was reported and the admin review

Signed-off-by: Lukas Matt <lukas@zauberstuhl.de>

jhass commented on app/models/report.rb:
```diff
@@ -14,6 +16,18 @@ class Report < ActiveRecord::Base
+    elsif item_type == COMMENT
+      Comment.find_by(id: item_id)
+    end
+  end
```
Adding an association to the `Report` model didn't work?

Zauberstuhl replied:
I am not quite sure how I can do that.

I could use something like [constantize](http://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-constantize) to get rid of `POST` and `COMMENT`

jhass replied:
Rails will do the magic for you if `item_type` contains the right thing: http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#module-ActiveRecord::Associations::ClassMethods-label-Polymorphic+Associations

jhass commented:
This broke a couple of specs https://travis-ci.org/diaspora/diaspora/jobs/80281877#L687

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+    @post = @user.post(:status_message, text: "hello", to: @user.aspects.first.id)
+    @comment = @user.comment!(@post, "welcome")
+
+    @post_report = @user.reports.create({
```
Redundant curly braces around a hash parameter.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+    @comment = @user.comment!(@post, "welcome")
+
+    @post_report = @user.reports.create({
+      :item_id => @post.id,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+
+    @post_report = @user.reports.create({
+      :item_id => @post.id,
+      :item_type => 'post',
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+    @post_report = @user.reports.create({
+      :item_id => @post.id,
+      :item_type => 'post',
+      :text => 'offensive content'
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+      :item_id => @post.id,
+      :item_type => 'post',
+      :text => 'offensive content'
+    })
```
Indent the right brace the same as the first position after the preceding left parenthesis.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+      :item_type => 'post',
+      :text => 'offensive content'
+    })
+    @comment_report = @user.reports.create({
```
Redundant curly braces around a hash parameter.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+      :text => 'offensive content'
+    })
+    @comment_report = @user.reports.create({
+      :item_id => @comment.id,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+    })
+    @comment_report = @user.reports.create({
+      :item_id => @comment.id,
+      :item_type => 'comment',
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+    @comment_report = @user.reports.create({
+      :item_id => @comment.id,
+      :item_type => 'comment',
+      :text => 'offensive content'
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
+      :item_id => @comment.id,
+      :item_type => 'comment',
+      :text => 'offensive content'
+    })
```
Indent the right brace the same as the first position after the preceding left parenthesis.

diaspora-code-review commented on spec/helpers/report_helper_spec.rb:
```diff
@@ -1,17 +1,29 @@
-    it "contains an anchor to the comment" do 
-      expect(helper.report_content(@comment, 'comment')).to include %Q(href="#{post_path(@post, anchor: @comment.guid)}")
+    it "contains an anchor to the comment" do
+      expect(helper.report_content(@comment_report)).to include %(href="#{post_path(@post, anchor: @comment.author.guid)}")
```
Line is too long. [123/120]

Commit:
Fix spec files and report model

* Adopt pronto suggestions

jhass commented on app/controllers/report_controller.rb:
```diff
@@ -7,7 +7,10 @@ class ReportController < ApplicationController
+    @reports ||= []
+    Report.where(reviewed: false).each do |report|
+      @reports << report unless report.item.nil?
+    end
```
Can that ever happen? how? anyway, should be a `select` if it's truly needed.

Zauberstuhl replied:
if the author deletes the post on its own after the report was made.
On second thought it might be better deleting the report on a post/comment-destroy/retract

jhass replied:
I guess so, yeah. Or make everything immune to a missing item.

jhass commented on app/models/report.rb:
```diff
@@ -14,6 +16,18 @@ class Report < ActiveRecord::Base
+  end
+
+  def reported_author
+    item.author unless item.nil?
```
`if item`

Commit:
Display all reports and give the ability to review them

Zauberstuhl commented:
Instead of filtering the admin could mark them as "reviewed" as well: 
![untitled](https://cloud.githubusercontent.com/assets/652428/10262708/a7d00bbc-69d3-11e5-8c9e-b528e5c275dd.png)

Objections?

jhass commented:
:+1:

diaspora-code-review commented on app/models/report.rb:
```diff
@@ -1,8 +1,8 @@
   validates :item_id, presence: true
-  validates :item_type, presence: true, :inclusion => { :in => %w(post comment),
-    :message => 'Type should match `post` or `comment`!'}
+  validates :item_type, presence: true, :inclusion => { :in => %w(Post Comment),
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Space inside { detected.

diaspora-code-review replied:
Align the elements of a hash literal if they span more than one line.

diaspora-code-review commented on app/models/report.rb:
```diff
@@ -1,8 +1,8 @@
-  validates :item_type, presence: true, :inclusion => { :in => %w(post comment),
-    :message => 'Type should match `post` or `comment`!'}
+  validates :item_type, presence: true, :inclusion => { :in => %w(Post Comment),
+    :message => 'Type should match `Post` or `Comment`!'}
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/models/report_spec.rb:
```diff
@@ -13,12 +13,12 @@
     @valid_post_report = {
       :item_id => @bob_post.id,
-      :item_type => 'post',
+      :item_type => 'Post',
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/models/report_spec.rb:
```diff
@@ -13,12 +13,12 @@
     @valid_comment_report = {
       :item_id => @bob_comment.id,
-      :item_type => 'comment',
+      :item_type => 'Comment',
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

Commit:
Use polymorphic association for the report item

* Adopt pronto suggestions

Signed-off-by: Lukas Matt <lukas@zauberstuhl.de>

jhass commented:
Thank you!

Flaburgan commented:
Thanks!
